subsys/bluetooth/services/ots/ots_dir_list.c

```diff
@@ -220,26 +191,72 @@ void bt_ots_dir_list_init(struct bt_ots_dir_list **dir_list, void *obj_manager)
 
 	memset(&dir_list_obj->metadata, 0, sizeof(dir_list_obj->metadata));
 	dir_list_obj->metadata.name = dir_list_obj_name;
-	dir_list_obj->metadata.size.alloc = sizeof((*dir_list)->_content);
+	dir_list_obj->metadata.size.alloc = DIR_LIST_MAX_SIZE;
 	dir_list_obj->metadata.type.uuid.type = BT_UUID_TYPE_16;
 	dir_list_obj->metadata.type.uuid_16.val = BT_UUID_OTS_DIRECTORY_LISTING_VAL;
 	BT_OTS_OBJ_SET_PROP_READ(dir_list_obj->metadata.props);
 
 	(*dir_list)->dir_list_obj = dir_list_obj;
 
-	/* Set size in dir_list_encode */
-	dir_list_encode(*dir_list, obj_manager);
+	dir_list_update_size(*dir_list, obj_manager);
 }
 
-ssize_t bt_ots_dir_list_content_get(struct bt_ots_dir_list *dir_list, void **data,
-				size_t len, off_t offset)
+ssize_t bt_ots_dir_list_content_get(struct bt_ots_dir_list *dir_list, void *obj_manager,
+				    void **data, size_t len, off_t offset)
 {
-	if (offset >= dir_list->net_buf.len) {
-		*data = NULL;
-		return 0;
-	}
+	struct bt_gatt_ots_object *obj;
+	int err;
+	size_t buffer_offset = 0;
+	size_t rec_len = 0;
+	size_t room = 0;
+	bool found = false;
+
+	err = bt_gatt_ots_obj_manager_first_obj_get(obj_manager, &obj);
+
+	__ASSERT(err == 0 && obj == dir_list->dir_list_obj,
+		 "Expecting first object to be the Directory Listing Object");
+
+	do {
+		/* keep resetting the buffer until we find the first object record that
+		 * is of intereset.
+		 */
+		if (!found) {
+			buffer_offset += rec_len;
+
+			net_buf_simple_init_with_data(&dir_list->net_buf,
+						      dir_list->_content,
+						      sizeof(dir_list->_content));
+			net_buf_simple_reset(&dir_list->net_buf);
+			room = sizeof(dir_list->_content);
+		}
+
+		rec_len = dir_list_object_encode(obj, &dir_list->net_buf);
 
-	*data = dir_list->net_buf.data + offset;
+		/* if this record contains at lease one byte that the client is asking
+		 * for then we have found our first entry and we should stop resetting
+		 * the buffer. We could only keep the bytes that are off intereset to
+		 * the client but then we would have a buffer that isn't a multiple of
+		 * object records and that would make things more complicated.
+		 */
+		found = buffer_offset + rec_len > offset;
+		room -= rec_len;
+
+		err = bt_gatt_ots_obj_manager_next_obj_get(obj_manager, obj, &obj);
+
+		/* keep accumulating object records while we haven't run out of objects
+		 * or room to write a full object record.
+		 */
+	} while (!err && room >= DIR_LIST_OBJ_RECORD_MAX_SIZE);
 
-	return MIN(len, dir_list->net_buf.len - offset);
+	/* since we keep full object records in the buffer, the buffer may contain data
+	 * that the client already has, so we must offset into the buffer.
+	 */
+	*data = dir_list->net_buf.data + (offset - buffer_offset);
+
+	return MIN(len, dir_list->net_buf.len - (offset - buffer_offset));
```

Comment:
If the buffer already contains data, then we shouldn't even start encoding new ones before the previous buffer has been sent.

Reply:
I am not sure I agree with this. The client doesn't have to read the rest of the data in the buffer. Also when it does it will ask for a new offset.